pkg/maps/ctmap/ctmap.go

```diff
@@ -513,95 +412,84 @@ func purgeCtEntry4(m *Map, key CtKey, entry *CtEntry, natMap *nat.Map, next func
+	if ACT != nil {
+		countFailedFn = ACT.CountFailed4
+		if ipv6 {
+			countFailedFn = ACT.CountFailed6
 		}
 	}
-
-	filterCallback := func(key bpf.MapKey, value bpf.MapValue) {
+	return func(key bpf.MapKey, value bpf.MapValue) {
+		// TODO: These type assertions are a bit dangerous, make more of this well typed
+		// to avoid having to make these assertions.
+		tupleKey := key.(tupleKeyAccessor)
+		ctKey := key.(CtKey)
 		entry := value.(*CtEntry)
 
-		switch obj := key.(type) {
-		case *CtKey4Global:
-			currentKey4Global := obj
-			// In CT entries, the source address of the conntrack entry (`SourceAddr`) is
-			// the destination of the packet received, therefore it's the packet's
-			// destination IP
-			action := filter.doFiltering(currentKey4Global.DestAddr.Addr(), currentKey4Global.SourceAddr.Addr(),
-				currentKey4Global.DestPort, currentKey4Global.SourcePort,
-				uint8(currentKey4Global.NextHeader), currentKey4Global.Flags, entry)
-
-			switch action {
-			case deleteEntry:
-				err := purgeCtEntry4(m, currentKey4Global, entry, natMap, next)
-				if err != nil {
-					log.WithError(err).WithField(logfields.Key, currentKey4Global.String()).Error("Unable to delete CT entry")
-				} else {
-					stats.deleted++
-				}
-			default:
-				stats.aliveEntries++
-			}
-		case *CtKey4:
-			currentKey4 := obj
-			// In CT entries, the source address of the conntrack entry (`SourceAddr`) is
-			// the destination of the packet received, therefore it's the packet's
-			// destination IP
-			action := filter.doFiltering(currentKey4.DestAddr.Addr(), currentKey4.SourceAddr.Addr(),
-				currentKey4.DestPort, currentKey4.SourcePort,
-				uint8(currentKey4.NextHeader), currentKey4.Flags, entry)
-
-			switch action {
-			case deleteEntry:
-				err := purgeCtEntry4(m, currentKey4, entry, natMap, next)
-				if err != nil {
-					log.WithError(err).WithField(logfields.Key, currentKey4.String()).Error("Unable to delete CT entry")
-				} else {
-					stats.deleted++
-				}
-			default:
-				stats.aliveEntries++
+		// In CT entries, the source address of the conntrack entry (`SourceAddr`) is
+		// the destination of the packet received, therefore it's the packet's
+		// destination IP
+		action := filter.doFiltering(tupleKey.GetDestAddr(), tupleKey.GetSourceAddr(),
+			tupleKey.GetDestPort(), tupleKey.GetSourcePort(),
+			uint8(tupleKey.GetNextHeader()), tupleKey.GetFlags(), entry)
+
+		switch action {
+		case deleteEntry:
+			err := purgeCtEntry(m, ctKey, entry, natMap, next, countFailedFn)
+			if err != nil {
+				log.WithError(err).WithField(logfields.Key, ctKey.String()).Error("Unable to delete CT entry")
+			} else {
+				stats.deleted++
 			}
 		default:
-			log.Warningf("Encountered unknown type while scanning conntrack table: %v", reflect.TypeOf(key))
+			stats.aliveEntries++
 		}
-	}
 
-	// We serialize the deletions in order to avoid forced map walk restarts
-	// when keys are being evicted underneath us from concurrent goroutines.
```

Comment:
Shouldn't we leave this comment? I think we still need the globalDeleteLock to serialize the deletions otherwise (*Endpoint).scrubIPsInConntrackTableLocked() removes the CT entries concurrently.

Reply:
Good point, ill read. Also expanded the comment a bit since I missed the meaning the first time I read it.